docs/reference/certificaterequests.rst

```diff
@@ -0,0 +1,77 @@
+     isCA: false
+     duraton: 90d
+     issuerRef:
+       name: ca-issuer
+       # We can reference ClusterIssuers by changing the kind here.
+       # The default value is Issuer (i.e. a locally namespaced Issuer)
+       kind: Issuer
+       group: certmanager.k8s.io
+
+This CertificateRequest will make cert-manager attempt to make the Issuer
+``letsencrypt-prod`` in the default issuer pool ``certmanager.k8s.io``, return a
+certificate based upon the certificate signing request. Other groups can be
+specified inside the ``issuerRef`` which will change the targeted issuers to other
+external, third party issuers you may have installed.
+
+The resource also exposes the option for stating the certificate as CA and
+requested validity duration.
+
+A successful issuance of the certificate signing request will cause an update to
+the resource, setting the status with the signed certificate, the CA of the
+certificate (if available), and setting the `Ready` condition to `True`.
+
+Whether issuance of the controller was successful or not, a retry of the
+issuance will _not_ happen. It is the responsibility of some other controller to
+manage the logic and life cycle of CertificateRequets.
+
+----------
+Conditions
+----------
+
+CertificateRequests have a set of strongly defined conditions that should be
+used and relied upon by controllers or services to make decisions on what
+actions to take next on the resource. Each condition consists of the pair
+`Ready` - a boolean value, and `Reason` - a string. The set of values and
+meanings are as follows:
+
++---------+-----------------+---------------------------------------------------------------------------------------------------------------------------+
+| *Ready* | *Reason*        | Condition Meaning                                                                                                         |
++=========+=================+===========================================================================================================================+
+| False   | CertPending     | A valid Issuer has been found and is currently in the process of issuing a certificate based on the Spec of the resource. |
++---------+-----------------+---------------------------------------------------------------------------------------------------------------------------+
+| False   | CertFailed      | The certificate has failed to be issued. No further action will be taken on the CertificateRequest by it's controller.    |
++---------+-----------------+---------------------------------------------------------------------------------------------------------------------------+
+| False   | IssuerNotFound  | The referenced Issuer detailed in the Spec cannot be found.                                                               |
```

Comment:
What is the expected action on the user's/controller's part in this state? Should it just wait? Should it delete the CertificateRequest? I think if we are going to define these as 'strongly defined states', we need to also advise what the expected behaviour is *in these states*.

I wonder if the IssuerInitError and IssuerNotFound reasons can be removed in favour of `Pending` and/or `Failed` with a suitable human readable 'Message' describing why?

esp. given I cannot currently think of an instance where it's useful to *programatically* know that specifically the issuer needs changing. If we use the message, I'd guess that the Certificate's controller could then display that message to the user in order for the issue to be remediated.

Reply:
Yep I think that makes sense to me too.

An issuer not being found should make the controller sit and wait - `Pending` seems more suitable.

For InitError - I feel like this is really a failed state. We've deemed that the issuer exists but it failed to construct an issuer to sign. This seems like more action should be taken by manual intervention. The error here does get bubbled up to the status.

Again, I'm going to make these change to code in a separate PR.

Reply:
👍